File: cobalt/backend/domain/repositories/servers.py
#  Copyright (C) 2026 ArtoriasCode
#  Author: ArtoriasCode
#  Repository: https://github.com/ArtoriasCode/cobalt
#  SPDX-License-Identifier: AGPL-3.0-or-later

from abc import ABC, abstractmethod
from typing import Optional, List, Any

from domain.entities import (
    ServerEntity,
    ServersPageEntity,
    ServersGetPageEntity,
    ServerCreateEntity,
    ServerUpdateEntity
)


class AbstractServersRepository(ABC):
    """
    Abstract servers repository.
    """

    @abstractmethod
    async def get_page(
        self,
        entity: ServersGetPageEntity,
        session: Optional[Any] = None
    ) -> ServersPageEntity:
        """
        Gets a paginated list of servers.

        Parameters:
        - entity: ServersGetPageEntity object.
        - session: Session object.

        Returns:
        - ServersPageEntity: ServersPageEntity object.
        """
        ...

    @abstractmethod
    async def get_one_by_id(
        self,
        server_id: int,
        session: Optional[Any] = None
    ) -> Optional[ServerEntity]:
        """
        Gets an existing server by ID.

        Parameters:
        - server_id: Server ID.
        - session: Session object.

        Returns:
        - ServerEntity: ServerEntity object.
        """
        ...

    @abstractmethod
    async def get_many_by_ids(
        self,
        server_ids: List[int],
        session: Optional[Any] = None
    ) -> List[ServerEntity]:
        """
        Gets multiple existing servers by IDs.

        Parameters:
        - server_ids: List of server IDs.
        - session: Session object.

        Returns:
        - List: List of ServerEntity objects.
        """
        ...

    @abstractmethod
    async def create_one(
        self,
        entity: ServerCreateEntity,
        session: Optional[Any] = None
    ) -> ServerEntity:
        """
        Creates a new server.

        Parameters:
        - entity: ServerCreateEntity object.
        - session: Session object.

        Returns:
        - ServerEntity: ServerEntity object.
        """
        ...

    @abstractmethod
    async def update_one(
        self,
        entity: ServerUpdateEntity,
        session: Optional[Any] = None
    ) -> Optional[ServerEntity]:
        """
        Updates an existing server.

        Parameters:
        - entity: ServerUpdateEntity object.
        - session: Session object.

        Returns:
        - ServerEntity: ServerEntity object.
        """
        ...

    @abstractmethod
    async def delete_one(
        self,
        server_id: int,
        session: Optional[Any] = None
    ) -> Optional[ServerEntity]:
        """
        Deletes an existing server.

        Parameters:
        - server_id: Server ID.
        - session: Session object.

        Returns:
        - ServerEntity: ServerEntity object.
        """
        ...

    @abstractmethod
    async def delete_many(
        self,
        server_ids: List[int],
        session: Optional[Any] = None
    ) -> List[ServerEntity]:
        """
        Deletes multiple existing servers.

        Parameters:
        - server_ids: List of server IDs.
        - session: Session object.

        Returns:
        - List: List of ServerEntity objects.
        """
        ...

File: cobalt/frontend/src/contracts/services/api/http/servers.ts
/*
 * Copyright (C) 2026 Artorias
 * Author: Artorias
 * Repository: https://github.com/artorias-developer/cobalt
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type {
  ServerEntity,
  ServersPageEntity,
  ServerStatusEntity,
  ServersPageRequest,
  ServerCreateRequest,
  ServerUpdateRequest,
  ServerExecuteRequest
} from "@/types"

/**
 * Interface for servers API service operations.
 * Defines contract for managing game servers.
 * Implementation-agnostic - not tied to any specific HTTP client.
 */
export interface IHttpServersApiService {
  /**
   * Gets a paginated list of servers.
   *
   * Parameters:
   * - params: ServersPageRequest object.
   *
   * Returns:
   * - Promise<ServersPageResponse>: Paginated list of servers.
   */
  getPage(params: ServersPageRequest): Promise<ServersPageEntity>

  /**
   * Gets an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<ServerEntity>: Server entity.
   */
  getOne(serverId: number): Promise<ServerEntity>

  /**
   * Creates a new server.
   *
   * Parameters:
   * - data: ServerCreateRequest object.
   *
   * Returns:
   * - Promise<ServerEntity>: Created server entity.
   */
  createOne(data: ServerCreateRequest): Promise<ServerEntity>

  /**
   * Updates an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   * - data: ServerUpdateRequest object.
   *
   * Returns:
   * - Promise<ServerEntity>: Updated server entity.
   */
  updateOne(serverId: number, data: ServerUpdateRequest): Promise<ServerEntity>

  /**
   * Deletes an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  deleteOne(serverId: number): Promise<void>

  /**
   * Deletes multiple existing servers.
   *
   * Parameters:
   * - ids: List of server IDs.
   *
   * Returns:
   * - Promise<void>.
   */
  deleteMany(ids: number[]): Promise<void>

  /**
   * Starts an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  start(serverId: number): Promise<void>

  /**
   * Stops an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  stop(serverId: number): Promise<void>

  /**
   * Restarts an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  restart(serverId: number): Promise<void>

  /**
   * Executes a command inside the server container.
   *
   * Parameters:
   * - serverId: Server ID.
   * - data: ServerExecuteRequest object.
   *
   * Returns:
   * - Promise<void>.
   */
  execute(serverId: number, data: ServerExecuteRequest): Promise<void>

  /**
   * Gets the server container status.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<ServerStatusEntity>: Server status entity.
   */
  status(serverId: number): Promise<ServerStatusEntity>
}